Avoid encoding console input on Windows when using Terminal.reader()

Minecrell opened the pull request:
On Windows, console input is read using the Windows API as UTF-16 characters. When using the provided Terminal.reader(), input is encoded using UTF-8 just to decode it again shortly after.

Similar to the changes made in #168, change Terminal.reader() to consume the characters directly to avoid unnecessary encoding.

Instead of a `PipedInput/OutputStream` a new `PumpReader` is now used that works very much like a `PipedReader/Writer`. However, it additionally supports reading input using an `InputStream` with a specific `Charset`. It's much easier and more reliable to implement this directly in the `Reader` because you need to buffer to encode characters correctly.

gnodet commented:
Looks good to me.  Let me know if you want to merge, but I guess this may be part of #174.

Minecrell commented:
@gnodet I have some changes locally to optimize console reading a bit, will push them tomorrow. We then need to merge this for #174 so JLine's `LineReader` works without being affected by the selected encoding.

Minecrell commented:
@gnodet The PR should be good to go now.

I've added 2 commits for some minor optimizations, one commit to avoid memory allocations each time something is read/written to the console, and another one to avoid buffering the console input in additional `StringBuilder`s (it's just written directly to the pump buffer instead). It should still behave the same, I haven't changed the way the input is processed.

I've also added a unit test for `PumpReader` to ensure it's working correctly and discovered two minor bugs in it. One of them is similar to https://github.com/jline/jline3/commit/0fa03a94d0e9481a4e09b849c0c1831db5610d09, reading wide chars was broken if the bytes were read separately. This is also fixed now.
